orttraining/orttraining/core/graph/training_op_defs.cc

```diff
@@ -3123,38 +3117,12 @@ Return true if all elements are true and false otherwise.
           {"tensor(int64)"},
           "Constrain input type to 64-bit integer.")
       .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
-        // Load expected input types.
-        const auto input_tensor_types_proto = ctx.getAttribute("input_tensor_types");
-        // This is a required field.
-        ORT_ENFORCE(input_tensor_types_proto, "PythonOpGrad's must have \"input_tensor_types\" attribute.");
-        // Check if the inferred input types match those described in the
-        // "input_tensor_types" attributes.
-        const auto input_tensor_requires_grads = ctx.getAttribute("input_tensor_requires_grads");
-        // Expected input schema: [ctx, grad_input_1, ..., grad_input_N, unused_1, ..., unused_M]
-        // "unused" inputs are just control inputs and they are not used actual computation.
-        // Other variables are used to invoke autograd.Function.backward(ctx, grad_input1, ..., grad_input_N).
-        // The "input_count" here means 1 + N.
-        const auto input_count = input_tensor_requires_grads->ints().size();
-        ORT_ENFORCE(input_tensor_types_proto->ints_size() == input_count - 1,
-                    "PythonOp's input list should have one more element than \"input_tensor_types\" attribute.");
-        // The first input is a pointer which points to
-        // a Python object created by torch.autograd.Function.apply.
-        // For details, see how we interpret it in PythonOpGrad implementation.
-        for (auto i = 1; i < input_count; ++i) {
```

Comment:
This doesn't look robust. I prefer to keep as many type checks as possible.

Reply:
I think it is fine, because this is the shape inference of PythonOpGrad, all those attributes for example input_tensor_requires_grads comes from PythonOp's attribute. When we create PythonOpGrad, we already do the check. Does that make sense?

Reply:
As we talked, I keep those attributes as redundancy, for more checks. Add some explanations in the op definitions in case of confusion of anyone curious why they are needed. :)